### ISSUE TITLE
union type checking gets wrong with readonly 

### ISSUE THREAD

**Issue by bysdxt** [microsoft/TypeScript#34949]
**TypeScript Version:**  3.7.2

**Code**
type A = { readonly [0]: 0 }
type B = { [0]: number }
declare var x: A | B;
if (x[0]) x[0] = 2;

**Expected behavior:**
No error will be reported

**Actual behavior:**
error TS2540: Cannot assign to '[0]' because it is a read-only property.

4 if (x[0]) x[0] = 2;
              ~

**Comment by jcalz** [comment]
The [new issue template](https://github.com/microsoft/TypeScript/blob/1c42c1aaa82de92d40030ae9e3ecb95831ad18f5/.github/ISSUE_TEMPLATE/Bug_report.md) asks you to fill out search terms and related issues, to help put your issue in context and show some effort on your part to ensure that the issue has not already been addressed.

I think this is a duplicate of #31045; you want `A | B` to be a discriminated union, but `number` is too wide to be considered a discriminant, and `0 | number` is just `number`.   So checking `x[0]` doesn't  eliminating `A` from the union, which is the issue.  If you change the type of `B`'s property to be a union of literals not including `0`, such as  `type B = {0: 1 | 2 | 3}`, the problem goes away.

**Comment by bysdxt** [comment]
> The [new issue template](https://github.com/microsoft/TypeScript/blob/1c42c1aaa82de92d40030ae9e3ecb95831ad18f5/.github/ISSUE_TEMPLATE/Bug_report.md) asks you to fill out search terms and related issues, to help put your issue in context and show some effort on your part to ensure that the issue has not already been addressed.
> 
> I think this is a duplicate of #31045; you want `A | B` to be a discriminated union, but `number` is too wide to be considered a discriminant, and `0 | number` is just `number`. So checking `x[0]` doesn't eliminate `A` from the union, which is the issue. If you change the type of `B`'s property to be a union of literals not including `0`, such as `type B = {0: 1 | 2 | 3}`, the problem goes away.

Think you.I will fill out search terms next time.
And you say "If you change the type of `B`'s property to be a union of literals not including `0`",but the code: `
type A = { readonly [0]: 0 }
type B = { [0]: -1 | 0 | 1 | 2 | 3 }
declare var x: A | B;
if (x[0]) x[0] = 2;
`
works, even  `B`'s property includes `0`

**Comment by jcalz** [comment]
You're right about that, sorry.  It's only important that the discriminant type not include `number` if it also includes numeric literal types.

**Comment by andrewbranch** [comment]
Note: this isn’t actually related to `readonly`. @jcalz is right about the root cause, and I _think_ is essentially asking the same thing as #31045, but this example is simpler. Here’s a snippet that I think shows the behavior in question without the distraction of `readonly`, one example with numbers and another with strings:

```ts
declare var n: { x: 0 } | { x: number, y: string };
if (n.x) {
    n.y;
}

declare var s: { x: "" } | { x: string, y: string };
if (s.x) {
    s.y;
}
```

https://www.typescriptlang.org/play/#code/CYUwxgNghgTiAEA3W8B2AueBveAPTADPAL7wA+2emqArgLYBGIMANPAJ6YDOALjAJaoA5iQDcAKH4AzeAApUAOlwBKbOPga0C9hOLjxoSLATIY8Lphz54AIhslyla7wHC2nc30EjiE6XK4lVSx1TUCdcT0gA
